File: 3.无重复字符的最长子串.cs
/*
 * @lc app=leetcode.cn id=3 lang=csharp
 *
 * [3] 无重复字符的最长子串
 */
using System.Collections.Generic;
// @lc code=start
public class Solution {
    public int LengthOfLongestSubstring(string s) {
        int left=0,right=0;
        int longest = 0;
        HashSet<char> chars = new HashSet<char>();

        while(right<s.Length)
        {
            if(chars.Contains(s[right])) {
                chars.Remove(s[left]);
                left++;
            }else{
                chars.Add(s[right]);
                right++;
                longest = longest>(chars.Count)?longest:chars.Count;
            }
        }

        return longest;
    }
}
// @lc code=end

File: py/3.无重复字符的最长子串.py
#
# @lc app=leetcode.cn id=3 lang=python3
#
# [3] 无重复字符的最长子串
#
# https://leetcode-cn.com/problems/longest-substring-without-repeating-characters/description/
#
# algorithms
# Medium (35.23%)
# Likes:    4085
# Dislikes: 0
# Total Accepted:    595.5K
# Total Submissions: 1.7M
# Testcase Example:  '"abcabcbb"'
#
# 给定一个字符串，请你找出其中不含有重复字符的 最长子串 的长度。
# 
# 示例 1:
# 
# 输入: "abcabcbb"
# 输出: 3 
# 解释: 因为无重复字符的最长子串是 "abc"，所以其长度为 3。
# 
# 
# 示例 2:
# 
# 输入: "bbbbb"
# 输出: 1
# 解释: 因为无重复字符的最长子串是 "b"，所以其长度为 1。
# 
# 
# 示例 3:
# 
# 输入: "pwwkew"
# 输出: 3
# 解释: 因为无重复字符的最长子串是 "wke"，所以其长度为 3。
# 请注意，你的答案必须是 子串 的长度，"pwke" 是一个子序列，不是子串。
# 
# 
#

# @lc code=start
class Solution:
    def lengthOfLongestSubstring(self, s: str) -> int:
        chars = set()
        left, right = 0, 0
        cur, maxi = 0, 0

        while right < len(s):
            if s[right] not in chars:
                chars.add(s[right])
                cur += 1
                maxi = max(cur, maxi)

                right += 1

            else:
                chars.remove(s[left])
                left += 1
                cur -= 1

        return maxi
# @lc code=end

